### PR TITLE
feat(site): Duplicate template

### LINKED ISSUE #6118: template editor: duplicate an existing template
I'd love to duplicate a template from the template editor.

I was thinking when you click “duplicate template,” it takes you to the “create template” page with everything pre-filled, including variables. 

The “create template” page will prompt for variables anyways so my thinking is it would be more beneficial than harmful to pre-fill those values

Because the user can always edit those fields.

And the “create template page” would also let the user quickly change the name from Kubernetes (copy) to something else without extra clicks

Open to other thoughts though

### PR DESCRIPTION
**Pull request by BrunoQuaresma**
https://user-images.githubusercontent.com/3165839/228310338-0d7cb0df-8d25-42bd-86b6-f18ed124a035.mov

I'm going to work on the tests when the flow is validated from @bpmct 

Closes #6118

### CODE REVIEW

**Review by rodrimaia — APPROVED** [review]

**Review by bpmct — DISMISSED** [review]
Awesome! Noticed a few things:

- Duplicating a template with no display name will make a display name ` Copy`
- Duplicating the built-in docker template returns the following error:
   
    https://user-images.githubusercontent.com/22407953/228349762-d7e33205-1114-4637-ae06-f1ab8da66536.mov

I tried importing another template with `coder templates init` (using Docker template) and ran into the same one. I tried with both `localhost:3000` and `localhost:8080`. Any idea why this is happening?

**Review by mtojek — COMMENTED** [review]
just a side note: do we need to check user template permissions here? so that only template authors can copy the template.

File: site/src/components/TemplateLayout/TemplatePageHeader.tsx [hunk 1]
```diff
@@ -28,6 +28,7 @@ const Language = {
   createButton: "Create workspace",
   deleteButton: "Delete",
   editFilesButton: "Edit files",
+  duplicateButton: "Duplicate",
```
[the remark this answers is not in the document]
Comment (BrunoQuaresma) [reply site/src/components/TemplateLayout/TemplatePageHeader.tsx:31]: In this case, the action is to literally duplicate so I think would make more sense than copy. 🤔

**Comment by BrunoQuaresma** [comment]
@mtojek Yes! Good catch.

**Comment by BrunoQuaresma** [comment]
@bpmct do you mind taking a look into it and see if that works as expected now pls?

File: site/src/components/TemplateLayout/TemplatePageHeader.tsx [hunk 2]
```diff
@@ -67,6 +68,15 @@ const TemplateMenu: FC<{
         >
           {Language.settingsButton}
         </MenuItem>
+        {canEditFiles && (
+          <MenuItem
+            onClick={handleClose}
+            component={RouterLink}
+            to={`/templates/new?copyTemplate=${templateName}`}
```
Comment (mtojek) [inline site/src/components/TemplateLayout/TemplatePageHeader.tsx:75]: maybe: `fromTemplate` or `from` ? `copyTemplate` might be confusing

File: site/src/pages/CreateTemplatePage/CreateTemplateForm.tsx [hunk 3]
```diff
@@ -146,7 +146,9 @@ const getInitialValues = ({
       ...initialValues,
       ...fromCopy,
       name: `${fromCopy.name}-copy`,
-      display_name: `${fromCopy.display_name} Copy`,
+      display_name: fromCopy.display_name
+        ? `${fromCopy.display_name} Copy`
```
Comment (mtojek) [inline site/src/pages/CreateTemplatePage/CreateTemplateForm.tsx:150]: nit: Copy of `${fromCopy.display_name}`?

File: site/src/xServices/createTemplate/createTemplateXService.ts [hunk 4]
```diff
@@ -388,6 +390,19 @@ export const createTemplateMachine =
             })
           }
 
+          if (templateNameToCopy) {
+            if (!version) {
+              throw new Error("Not able to copy without having a version")
```
Comment (mtojek) [inline site/src/xServices/createTemplate/createTemplateXService.ts:395]: nit: can't copy template due to a missing template version

File: site/src/xServices/createTemplate/createTemplateXService.ts [hunk 5]
```diff
@@ -61,6 +63,9 @@ interface CreateTemplateContext {
   // uploadedFile is the response from the server to use in the API
   file?: File
   uploadResponse?: UploadResponse
+  // When wanting to copy a Template
```
Comment (mtojek) [inline site/src/xServices/createTemplate/createTemplateXService.ts:66]: nit: `// Duplicate a template:`

**Review by mtojek — COMMENTED** [review]
The code logic is 👍 . My biggest concern is the lack of consistency between the public name and codebase: `Duplicate a template` vs. `copy`. I think that you should unify this, and just stick to `duplicate`. I know that these are synonyms, but it will be easier to work with code later.

**Pull request merged**